File: rust/src/properties.rs
use crate::types::*;

pub const PROPERTY_FUNCTIONS: [PropertyFn; crate::constants::NUMBER_OF_PROPERTIES] = [
    antisymmetry,
    antitransitivity,
    asymmetry,
    coreflexivity,
    density,
    irreflexivity,
    left_euclideanness,
    left_quasi_reflexivity,
    quasi_reflexivity,
    reflexivity,
    right_euclideanness,
    right_quasi_reflexivity,
    strict_density,
    symmetry,
    totality,
    transitivity,
    trichotomy,
];

macro_rules! get {
    ($relation:expr, $row:expr, $col:expr) => {
        ($relation[$row] >> $col) & 1 == 1
    };
}

fn antisymmetry(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in i + 1..set_size {
            if get!(relation, i, j) && get!(relation, j, i) {
                return false;
            }
        }
    }

    true
}

fn antitransitivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if relation[i] == 0 {
            continue;
        }

        for j in 0..set_size {
            if get!(relation, i, j) && (relation[i] & relation[j] != 0) {
                return false;
            }
        }
    }

    true
}

fn asymmetry(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if get!(relation, i, i) {
            return false;
        }

        for j in i + 1..set_size {
            if get!(relation, i, j) && get!(relation, j, i) {
                return false;
            }
        }
    }

    true
}

fn coreflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if relation[i] & !(1 << i) != 0 {
            return false;
        }
    }

    true
}

fn density(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            if !get!(relation, i, j) {
                continue;
            }

            let mut found = false;

            for k in 0..set_size {
                if get!(relation, i, k) && get!(relation, k, j) {
                    found = true;
                    break;
                }
            }

            if !found {
                return false;
            }
        }
    }

    true
}

fn irreflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if get!(relation, i, i) {
            return false;
        }
    }

    true
}

fn left_euclideanness(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            for k in 0..set_size {
                if get!(relation, j, i) && get!(relation, k, j) && !get!(relation, k, i) {
                    return false;
                }
            }
        }
    }

    true
}

fn left_quasi_reflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            if i != j && get!(relation, i, j) && !get!(relation, i, i) {
                return false;
            }
        }
    }

    true
}

fn quasi_reflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            if i != j && get!(relation, i, j) && !(get!(relation, i, i) && get!(relation, j, j)) {
                return false;
            }
        }
    }

    true
}

fn reflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if !get!(relation, i, i) {
            return false;
        }
    }

    true
}

fn right_euclideanness(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            for k in 0..set_size {
                if get!(relation, i, j) && get!(relation, i, k) && !get!(relation, j, k) {
                    return false;
                }
            }
        }
    }

    true
}

fn right_quasi_reflexivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            if i != j && get!(relation, i, j) && !get!(relation, j, j) {
                return false;
            }
        }
    }

    true
}

fn strict_density(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in 0..set_size {
            if i == j || !get!(relation, i, j) {
                continue;
            }

            let mut found = false;

            for k in 0..set_size {
                if k == i || k == j {
                    continue;
                }

                if get!(relation, i, k) && get!(relation, k, j) {
                    found = true;
                    break;
                }
            }

            if !found {
                return false;
            }
        }
    }

    true
}

fn symmetry(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in i + 1..set_size {
            if get!(relation, i, j) != get!(relation, j, i) {
                return false;
            }
        }
    }

    true
}

fn totality(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if !get!(relation, i, i) {
            return false;
        }

        for j in i + 1..set_size {
            if !(get!(relation, i, j) || get!(relation, j, i)) {
                return false;
            }
        }
    }

    true
}

fn transitivity(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        if relation[i] == 0 {
            continue;
        }

        for j in 0..set_size {
            if get!(relation, i, j) && (relation[j] & !relation[i] != 0) {
                return false;
            }
        }
    }

    true
}

fn trichotomy(relation: Relation, set_size: usize) -> bool {
    for i in 0..set_size {
        for j in i + 1..set_size {
            if !(get!(relation, i, j) || get!(relation, j, i)) {
                return false;
            }
        }
    }

    true
}

File: c/src/properties.c
#include "properties.h"

#include <stdbool.h>
#include <stdint.h>

#include "constants.h"

bool (*const PROPERTY_FUNCTIONS[NUMBER_OF_PROPERTIES])(uint8_t[], uint8_t) = {
    antisymmetry,
    antitransitivity,
    asymmetry,
    coreflexivity,
    density,
    irreflexivity,
    left_euclideanness,
    left_quasi_reflexivity,
    quasi_reflexivity,
    reflexivity,
    right_euclideanness,
    right_quasi_reflexivity,
    strict_density,
    symmetry,
    totality,
    transitivity,
    trichotomy};

bool antisymmetry(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = i + 1; j < set_size; j++)
            if (GET(relation, i, j) && GET(relation, j, i)) return false;

    return true;
}

bool antitransitivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (int i = 0; i < set_size; i++) {
        if (!relation[i]) continue;

        for (int j = 0; j < set_size; j++)
            if ((GET(relation, i, j)) && (relation[i] & relation[j]))
                return false;
    }

    return true;
}

bool asymmetry(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++) {
        if (GET(relation, i, i)) return false;

        for (uint8_t j = i + 1; j < set_size; j++)
            if (GET(relation, i, j) && GET(relation, j, i)) return false;
    }

    return true;
}

bool coreflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        if (relation[i] & ~(1 << i)) return false;

    return true;
}

bool density(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            if (GET(relation, i, j)) {
                bool found = false;

                for (uint8_t k = 0; k < set_size; k++) {
                    if (GET(relation, i, k) && GET(relation, k, j)) {
                        found = true;
                        break;
                    }
                }

                if (!found) return false;
            }

    return true;
}

bool irreflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        if (GET(relation, i, i)) return false;

    return true;
}

bool left_euclideanness(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            for (uint8_t k = 0; k < set_size; k++)
                if (GET(relation, j, i) && GET(relation, k, i) &&
                    !GET(relation, j, k))
                    return false;

    return true;
}

bool left_quasi_reflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            if (i != j && GET(relation, i, j) && !GET(relation, i, i))
                return false;

    return true;
}

bool quasi_reflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            if (i != j && GET(relation, i, j) &&
                !(GET(relation, i, i) && GET(relation, j, j)))
                return false;

    return true;
}

bool reflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        if (!GET(relation, i, i)) return false;

    return true;
}

bool right_euclideanness(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            for (uint8_t k = 0; k < set_size; k++)
                if (GET(relation, i, j) && GET(relation, i, k) &&
                    !GET(relation, j, k))
                    return false;

    return true;
}

bool right_quasi_reflexivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++)
            if (i != j && GET(relation, i, j) && !GET(relation, j, j))
                return false;

    return true;
}

bool strict_density(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = 0; j < set_size; j++) {
            if (i == j || !GET(relation, i, j)) continue;

            bool found = false;

            for (uint8_t k = 0; k < set_size; k++) {
                if (k == i || k == j) continue;
                if (GET(relation, i, k) && GET(relation, k, j)) {
                    found = true;
                    break;
                }
            }

            if (!found) return false;
        }

    return true;
}

bool symmetry(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = i + 1; j < set_size; j++)
            if (GET(relation, i, j) != GET(relation, j, i)) return false;

    return true;
}

bool totality(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++) {
        if (!GET(relation, i, i)) return false;

        for (uint8_t j = i + 1; j < set_size; j++)
            if (!(GET(relation, i, j) | GET(relation, j, i))) return false;
    }

    return true;
}

bool transitivity(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (int i = 0; i < set_size; i++) {
        if (!relation[i]) continue;

        for (int j = 0; j < set_size; j++)
            if ((GET(relation, i, j)) && (relation[j] & ~relation[i]))
                return false;
    }

    return true;
}

bool trichotomy(uint8_t relation[MAX_SET_SIZE], uint8_t set_size) {
    for (uint8_t i = 0; i < set_size; i++)
        for (uint8_t j = i + 1; j < set_size; j++)
            if (!(GET(relation, i, j) | GET(relation, j, i))) return false;

    return true;
}
